smallrye-reactive-messaging-kafka/src/test/java/io/smallrye/reactive/messaging/kafka/KafkaSourceConcurrencyTest.java

```diff
@@ -0,0 +1,322 @@
+import org.junit.jupiter.api.Test;
+
+import io.smallrye.reactive.messaging.kafka.base.KafkaTestBase;
+import io.smallrye.reactive.messaging.kafka.base.MapBasedConfig;
+import io.smallrye.reactive.messaging.kafka.impl.KafkaSource;
+
+/**
+ * Kafka Source Concurrency Experiment:
+ * <p>
+ * Hypothesis: If we match a topic's partition count with consumer instances
+ * then we should see concurrent record processing.
+ * <p>
+ * We will use topics with 2 partitions.
+ * We will send 2 records with keys "1" and "2".
+ * We will test that both records are processed and done so through separate partitions within 25 seconds.
+ * <p>
+ * For control we will provide a bean that will process both records as quickly as possible.
+ * For the experiment we will provide beans that purposefully sleep for 30 seconds after processing (prevent ack)
+ * <p>
+ * The idea here is that a force sleep should only delay the processing of records from 1 partition and not all.
+ */
+class KafkaSourceConcurrencyTest extends KafkaTestBase {
+
+    private static final int RECORD_COUNT = 2;
+
+    private interface ExperimentBean {
+        List<IncomingKafkaRecord<String, Integer>> getResults();
+    }
+
+    private static void waitFor30SecondsOrAllRecordsProcessed(ExperimentBean bean) {
+        for (int i = 0; i < 30; i++) {
+            if (bean.getResults().size() >= RECORD_COUNT) {
+                break;
+            }
+            try {
+                Thread.sleep(1_000L);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    /**
+     * Control bean
+     * <p>
+     * processes the record and ack as quickly as possible
+     */
+    @ApplicationScoped
+    public static class ControlBean implements ExperimentBean {
+
+        private final List<IncomingKafkaRecord<String, Integer>> list = new CopyOnWriteArrayList<>();
+
+        @Override
+        public List<IncomingKafkaRecord<String, Integer>> getResults() {
+            return list;
+        }
+
+        @Incoming("data")
+        public CompletionStage<Void> process(IncomingKafkaRecord<String, Integer> input) {
+            list.add(input);
+            return input.ack();
+        }
+    }
+
+    /**
+     * Experiment bean that uses {@link Acknowledgment.Strategy.POST_PROCESSING}
+     * <p>
+     * This tests an incoming channel with defaults
+     */
+    @ApplicationScoped
+    public static class PostAckWith30SecondSleepBean implements ExperimentBean {
+
+        private final List<IncomingKafkaRecord<String, Integer>> list = new CopyOnWriteArrayList<>();
+
+        @Override
+        public List<IncomingKafkaRecord<String, Integer>> getResults() {
+            return list;
+        }
+
+        @Incoming("data")
+        public CompletionStage<Void> process(IncomingKafkaRecord<String, Integer> input) {
+            list.add(input);
+            waitFor30SecondsOrAllRecordsProcessed(this);
```

Comment:
you are blocking the event loop, it's not legal.

Reply:
Right, but this us just an experiment. One of the attempts does not block the event loop (process in a worker thread) and the result is the same.